[dnf5] microdnf: return proper exit codes from commands

m-blaha opened the pull request.

m-blaha committed:
libdnf-cli: Exceptions structure for libdnf::cli

Patch introduces several libdnf::cli exceptions:
- AbortedByUserError for situations where user aborts the operation
  (e.g. by not confirming the transaction)
- GoalResolveError thrown when solver is not able to resolve the
  transaction
- CommandExitError when command needs to return non-standard exit code

m-blaha committed:
microdnf: Transaction commands return proper exit codes

lukash commented on include/libdnf-cli/exception.hpp:
```diff
@@ -0,0 +1,67 @@
+#include "libdnf/common/exception.hpp"
+
+namespace libdnf::cli {
+
```
For a proper exception hierarchy I think there should be a common base class `libdnf::cli::Error`. Not sure if it's practically needed but seems like a good idea formally.

m-blaha replied:
Added.

lukash commented on include/libdnf-cli/exception.hpp:
```diff
@@ -0,0 +1,67 @@
+
+namespace libdnf::cli {
+
+class CommandExecutionError : public Error {
```
The commands can still normally throw any exception as an error and we'll return the default error exit code, right? Should we call this class `CommandExitError`, make it not have a default exit code value, and only use this exception for the special cases where a different exit code is required? Presumably you'd also most of the time do a `throw_with_nested` as you'd already have an exception happening.

If you don't need a special exit code, you already likely have a specific exception class and you can throw it directly (and it's impractical to catch and `throw_with_nested` all exceptions everywhere if we wanted to throw this class the way you have it now for consistency).

lukash commented on include/libdnf-cli/exception.hpp:
```diff
@@ -0,0 +1,67 @@
+    explicit CommandExecutionError(int error_code = static_cast<int>(ExitCode::ERROR))
+        : Error("Command failed"),
+          error_code(error_code),
+          message_set(false) {}
```
I think this constructor is flat out bad :slightly_smiling_face: There should always be a message.

lukash commented on microdnf/commands/install/install.cpp:
```diff
@@ -87,16 +88,15 @@ void InstallCommand::run() {
     auto transaction = goal.resolve(false);
     if (transaction.get_problems() != libdnf::GoalProblem::NO_PROBLEM) {
-        return;
+        throw CommandExecutionError();
```
This is just bad. There should be, as usual, as complete information about the error in the message as possible. (I understand this will have to be amended in the future...)

lukash commented on microdnf/commands/install/install.cpp:
```diff
@@ -87,16 +88,15 @@ void InstallCommand::run() {
     if (!userconfirm(ctx.base.get_config())) {
-        std::cout << "Operation aborted." << std::endl;
-        return;
+        throw CommandExecutionError(1, "Operation aborted.");
```
I'd say _"Operation aborted by the user."_ just to be more clear, but this has the `1` constant which is the default error status, so this could use either just `libdnf::cli::Error` or a special error class `libdnf::cli::AbortedByUserError`?

I wonder if this exception (aborted by user) could ever be propagated through a plugin. If yes, it should definitely have its own class. If not, the class probably doesn't really matter.

m-blaha committed:
libdnf/transaction: Make get_resolve_logs() method const

m-blaha committed:
libdnf/transaction: Do not report resolvement problems to terminal

m-blaha committed:
libdnf-cli: ArgumentParserError inherits from libdnf::cli::Error

lukash commented on libdnf/base/transaction.cpp:
```diff
@@ -145,8 +144,7 @@ void Transaction::Impl::add_resolve_log(
     resolve_logs.emplace_back(LogEvent(problem, problems));
-    // TODO(jmracek) Use a logger properly and change a way how to report to terminal
-    std::cout << resolve_logs.back().to_string() << std::endl;
+    // TODO(jmracek) Use a logger properly
```
FTR I've been annoyed by these before, but I wasn't sure they can be removed without a proper replacement. (aren't they required in microdnf output?)

@j-mracek

m-blaha replied:
There is one thing still missing (I'll fix it in this PR) - resolve logs are not presented to the user in case the transaction is correctly resolved (ends with `libdnf::GoalProblem::NO_PROBLEM`). In case of problems they are printed via `GoalResolveError` exception.

m-blaha replied:
Fixed. Now the resolve logs are printed together with transaction table.

m-blaha committed:
libdnf/transaction: New API to get resolve logs as strings

Given that one of the most common use cases with resolve logs is
printing them, I'm adding this shortcut method.
It will also simplify unification of printing error logs acquired via
libdnf and D-Bus API.

m-blaha committed:
dnfdamon-server: Adapt to new get_resolve_logs_as_strings() API

m-blaha committed:
dnfdaemon-client: Add resolve_logs to DbusGoalWrapper

m-blaha committed:
libdnf-cli: Print resolve logs with transaction table

Even correctly resolved transaction can contain some warnings / hints /
infos in resolve logs (e.g. the package that user wanted to install is
already installed).
This patch prints those logs right before the transaction table.

m-blaha committed:
dnfdaemon-client: Unify transation table and exit codes with microdnf
